aioxmpp/pep/service.py

```diff
@@ -0,0 +1,238 @@
+
+    There also is a low-level interface:
+
+    .. automethod:: claim_pep_node
+
+    .. automethod:: unclaim_pep_node
+
+    """
+    ORDER_AFTER = [
+        aioxmpp.DiscoClient,
+        aioxmpp.DiscoServer,
+        aioxmpp.PubSubClient
+    ]
+
+    def __init__(self, client, **kwargs):
+        super().__init__(client, **kwargs)
+        self._pubsub = self.dependencies[aioxmpp.PubSubClient]
+        self._disco_client = self.dependencies[aioxmpp.DiscoClient]
+        self._disco_server = self.dependencies[aioxmpp.DiscoServer]
+
+        self._pep_node_claims = {}
+
+    def claim_pep_node(self, node_namespace, handler, *,
+                       register_feature=True, notify=False):
+        """
+        Claim node `node_namespace`.
+
+        Dispatch event notifications for `node_namespace` to `handler`.
+
+        This registers `node_namespace` as feature for service discovery
+        unless ``register_feature=False`` is passed.
+
+        :param node_namespace: the pubsub node whose events shall be
+            handled.
+        :param handler: the handler to install.
+        :type handler: callable, see
+            :attr:`aioxmpp.PubSubClient.on_item_publish`
+            for the arguments.
+        :param register_feature: Whether to publish the `node_namespace`
+            as feature.
+        :param notify: Whether to register the ``+notify`` feature to
+            receive notification without explicit subscription.
+
+        :raises RuntimeError: if a handler for `node_namespace` is already
+            set.
+        """
+        if node_namespace in self._pep_node_claims:
+            raise RuntimeError(
+                "setting handler for already handled namespace"
+            )
+        if register_feature:
+            self._disco_server.register_feature(node_namespace)
+        if notify:
+            self._disco_server.register_feature(node_namespace + "+notify")
+        self._pep_node_claims[node_namespace] = handler, register_feature, notify
+
+    def unclaim_pep_node(self, node_namespace):
+        """
+        Unclaim `node_namespace`.
+
+        The feature for `node_namespace` and the ``+notify`` feature
+        are automatically retracted if they were set by
+        :method:`claim_pep_node`.
+
+        :param node_namespace: The PubSub node whose handler shall be unset.
+
+        :raises KeyError: If the no handler is registered for
+            `node_namespace`.
+        """
+        _, feature, notify = self._pep_node_claims.pop(node_namespace)
+        if notify:
+            self._disco_server.unregister_feature(node_namespace + "+notify")
+        if feature:
+            self._disco_server.unregister_feature(node_namespace)
+
+    @asyncio.coroutine
+    def _check_for_pep(self):
+        # XXX: should this be done when the stream connects
+        # and we use the cached result later on (i.e. disable
+        # the PEP service if the server does not support PEP)
```

Comment:
Okay, I think the correct solution for now is to ignore this issue. Cases where PEP support vanishes during a reconnect should be *very* rare (as it requires a reconfigured server). We might want to add a call to check_for_pep when registering notify features, but even that isn't that important, we will simply not get notifications, and that's it. (And I do not see a clean solution that works in all cases).